rapids_logger/logger_impl.hpp.in

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../logger.hpp"
+
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/sinks/ostream_sink.h>
+#include <spdlog/spdlog.h>
+
+#include <memory>
+#include <sstream>
+#include <string>
+
+
+namespace __attribute__((visibility("hidden"))) @_RAPIDS_LOGGER_NAMESPACE@ {
+
+namespace detail {
+
+level_enum string_to_level(std::string_view const env_lvl_str)
+{
+    if (env_lvl_str == "TRACE") return level_enum::trace;
+    if (env_lvl_str == "DEBUG") return level_enum::debug;
+    if (env_lvl_str == "INFO") return level_enum::info;
+    if (env_lvl_str == "WARN") return level_enum::warn;
+    if (env_lvl_str == "ERROR") return level_enum::error;
+    if (env_lvl_str == "CRITICAL") return level_enum::critical;
+    if (env_lvl_str == "OFF") return level_enum::off;
+    std::ostringstream os{};
+    os << "Invalid logging level: " << env_lvl_str;
+    throw std::invalid_argument(os.str());
+}
+
+spdlog::level::level_enum to_spdlog_level(level_enum lvl)
+{
+  return static_cast<spdlog::level::level_enum>(static_cast<int32_t>(lvl));
+}
+
+level_enum from_spdlog_level(spdlog::level::level_enum lvl)
+{
+  return static_cast<level_enum>(static_cast<int32_t>(lvl));
+}
+
+}  // namespace detail
+
+
+// Sink methods
+sink::sink(spdlog::sink_ptr sink) : impl{sink} {}
+
+// Logger methods
+// TODO: This constructor and the one below introduce some undesirable
+// vtable/typeinfo public symbols for basic_file_sink and ostream_sink. I'm not
+// sure how to avoid this since they come from std::shared_ptr having inline
+// methods that require injection into the global symbol table for
+// disambiguation across TUs. I'm also not sure that we entirely need to
+// support this kind of constructor, so we might just want to remove it. At
+// minimum we will always have to expose the
+// std::shared_ptr<spdlog::sinks::sink> add_sink/remove_sink functions, but
+// those don't introduce any vtable or tbypeinfo entries (presumably because
+// the class doesn't actually implement any virtual functions).
```

Comment:
These are the symbols that we need to check before merging.

Reply:
I think I'll need to establish a full PImpl to really get rid of these. I'll look into that next.